src/runtime_strings_extra.js

```diff
@@ -62,16 +62,16 @@ function UTF16ToString(ptr, maxBytesToRead) {
   } else {
 #endif // TEXTDECODER != 2
 #endif // TEXTDECODER
-    var i = 0;
-
     var str = '';
-    while (1) {
+
+    for (var i = 0; !(i >= maxBytesToRead / 2); ++i) {
       var codeUnit = {{{ makeGetValue('ptr', 'i*2', 'i16') }}};
-      if (codeUnit == 0 || i == maxBytesToRead / 2) return str;
-      ++i;
+      if (codeUnit == 0) break;
       // fromCharCode constructs a character from a UTF-16 code unit, so we can pass the UTF16 string right through.
       str += String.fromCharCode(codeUnit);
     }
+
+    return str;
```

Comment:
Wait, how did this work at all without the return statement?   Was this simply not tested anywhere?

Reply:
The original while loop contained and if with an inconspicuous return. With null-terminated string it actually worked fine, but with non-null-terminated strings (and explicit length) it read one char past the end of the buffer.